airflow/www_rbac/security.py

```diff
@@ -120,60 +138,266 @@
+        sesh.commit()
+        if deleted_count:
+            logging.info('Deleted {} faulty permissions'.format(deleted_count))
+
+    def _merge_perm(self, permission_name, view_menu_name):
+        """
+        Add the new permission , view_menu to ab_permission_view_role if not exists.
+        It will add the related entry to ab_permission
+        and ab_view_menu two meta tables as well.
+
+        :param str permission_name: Name of the permission.
+        :param str view_menu_name: Name of the view-menu
+
+        :return:
+        """
+        permission = self.find_permission(permission_name)
+        view_menu = self.find_view_menu(view_menu_name)
+        pv = None
+        if permission and view_menu:
+            pv = self.get_session.query(self.permissionview_model).filter_by(
+                permission=permission, view_menu=view_menu).first()
+        if not pv and permission_name and view_menu_name:
+            self.add_permission_view_menu(permission_name, view_menu_name)
+
+    def create_custom_dag_permission_view(self):
+        """
+        Workflow:
+        1. when scheduler found a new dag, we will create an entry in ab_view_menu
+        2. we fetch all the roles associated with dag users.
+        3. we join and create all the entries for ab_permission_view_menu
+           (predefined permissions * dag-view_menus)
+        4. Create all the missing role-permission-views for the ab_role_permission_views
+
+        :return: None.
+        """
+        # todo(Tao): should we put this function here or in scheduler loop?
+        logging.info('Fetching a set of all permission, view_menu from FAB meta-table')
+
+        def merge_pv(perm, view_menu):
+            """Create permission view menu only if it doesn't exist"""
+            if view_menu and perm and (view_menu, perm) not in all_pvs:
+                self._merge_perm(perm, view_menu)
+
+        all_pvs = set()
+        for pv in self.get_session.query(self.permissionview_model).all():
+            if pv.permission and pv.view_menu:
+                all_pvs.add((pv.permission.name, pv.view_menu.name))
+
+        # create perm for global logical dag
+        for dag in dag_vms:
+            for perm in dag_perms:
+                merge_pv(perm, dag)
+
+        # Get all the active / paused dags and insert them into a set
+        all_dags_models = settings.Session.query(models.DagModel)\
+            .filter(or_(models.DagModel.is_active, models.DagModel.is_paused))\
+            .filter(~models.DagModel.is_subdag).all()
+
+        for dag in all_dags_models:
+            for perm in dag_perms:
+                merge_pv(perm, dag.dag_id)
+
+        # for all the dag-level role, add the permission of viewer
+        # with the dag view to ab_permission_view
+        all_roles = self.get_all_roles()
+        user_role = self.find_role('User')
+
+        dag_role = [role for role in all_roles if role.name not in EXISTING_ROLES]
```

Comment:
related to the other comments, this would become an issue if role names change. one way around this is to add a prefix to all dag_specific roles to differentiate them from user_created roles. alternatively we perhaps create our own Role model class and add a 'type' column that specifies whether it's a user role or a dag role.

Reply:
hey @jgao54 , same comment. I like the role model idea. But I hope we could do it in a later smaller pr.

Reply:
👍  not an issue if we make system roles ready-only :)